Add TFVisionEncoderDecoderModel

ydshieh opened the pull request:
# What does this PR do?

To make Vision-Encoder-Text-Decoder family complete by adding `TFVisionEncoderDecoderModel`.

To complete this PR, it requires to wait #13778 being merged to master (then rebase)
(And if we want to include a real integration test using the recent image-captioning ViT + GPT2 model, need to wait #14038 too)

ydshieh committed:
Start the work on TFVisionEncoderDecoderModel

ydshieh committed:
Expose TFVisionEncoderDecoderModel

ydshieh committed:
fix import

ydshieh committed:
Add modeling_tf_vision_encoder_decoder to _ignore_modules in get_model_modules()

ydshieh committed:
reorder

ydshieh committed:
Apply the fix for checkpoint loading as in #14016

ydshieh committed:
remove attention_mask + fix VISION_DUMMY_INPUTS

ydshieh committed:
A minimal change to make TF generate() work for vision models as encoder in encoder-decoder setting

ydshieh committed:
fix wrong condition: shape_list(input_ids) == 2

ydshieh committed:
add tests

ydshieh committed:
use personal TFViTModel checkpoint (for now)

ydshieh committed:
Add equivalence tests + projection layer

ydshieh committed:
style

ydshieh committed:
make sure projection layer can run

ydshieh committed:
Add examples

sgugger commented on src/transformers/generation_tf_utils.py:
```diff
@@ -614,14 +615,21 @@ def generate(
             input_ids = tf.fill((batch_size, 1), bos_token_id)
-        else:
-            assert len(shape_list(input_ids)) == 2, "Input prompt should be of shape (batch_size, sequence length)."
+        # TODO: remove this block
+        # else:
+        #     assert len(shape_list(input_ids)) == 2, "Input prompt should be of shape (batch_size, sequence length)."
```
Should be cleaned up?

sgugger commented on src/transformers/generation_tf_utils.py:
```diff
@@ -677,21 +685,35 @@ def generate(
+                "return_dict": return_dict_in_generate,
+            }
+
+            # vision models don't use `attention_mask`.
+            # PT's generate() also requires a similar change, at the following line
+            #   "if model_kwargs.get("attention_mask", None) is None:"
+            # TODO: change PT's generate() and remove `attention_mask` from `ViTModel`
```
Act on the TODO and remove the comment?

ydshieh replied:
Done

sgugger commented on src/transformers/generation_tf_utils.py:
```diff
@@ -677,21 +685,35 @@ def generate(
                 if output_hidden_states:
                     model_kwargs["encoder_hidden_states"] = encoder_outputs.hidden_states
 
+        # Add an extra condition `len(shape_list(input_ids)) == 2`, so this block will treat only text inputs.
+        # (vision inputs might occur when the model is an encoder-decoder model)
+        # In PT generate(), the `input_ids` is reset for decoder at the following line, which is before the expansion:
+        #   "# set input_ids as decoder_input_ids"
+        # TODO: it would be better to have the same approach as in PT's generate()
```
Reformulate the comment to just explain why there is a `len(shape_list(input_ids)) == 2` here.

sgugger commented on src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:
```diff
@@ -213,13 +213,6 @@ def dummy_inputs(self):
         # Add `decoder_input_ids` because `self.decoder` requires it.
         input_ids = tf.constant(DUMMY_INPUTS)
         dummy = {"input_ids": input_ids, "decoder_input_ids": input_ids}
-        # Add `encoder_hidden_states` to make the cross-attention layers' weights initialized
-        if self.config.add_cross_attention:
-            batch_size, seq_len = input_ids.shape
-            shape = (batch_size, seq_len) + (self.config.hidden_size,)
-            h = tf.random.uniform(shape=shape)
-            dummy["encoder_hidden_states"] = h
-
```
Why is this part removed?

ydshieh replied:
`TFEncoderDecoderModel.call()` doesn't have `encoder_hidden_states` parameter, but `encoder_outputs`.
Moreover, `encoder_hidden_states` is always passed to the decoder with `encoder_hidden_states = encoder_outputs[0]`. Therefore, there is no need to add `encoder_hidden_states` in `dummy_inputs`.

(I don't remember why I did that before, probably in some intermediate commits, it was required)

patrickvonplaten replied:
Looks good to me!

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,715 @@
+            }
+
+            # Add arguments to encoder from `kwargs_encoder`
+            for k, v in kwargs_encoder.items():
+                encoder_processing_inputs[k] = v
```
There is a function for this:
```suggestion
            encoder_processing_inputs.update(kwargs_encoder)
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,715 @@
+        }
+
+        # Add arguments to decoder from `kwargs_decoder`
+        for k, v in kwargs_decoder.items():
+            decoder_processing_inputs[k] = v
```
```suggestion
        decoder_processing_inputs.update(kwargs_decoder)
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,715 @@
+            cross_attentions=cross_attns,
+        )
+
+    def prepare_inputs_for_generation(
+        self,
+        decoder_input_ids,
+        past,
+        use_cache=None,
+        **kwargs,
+    ):
```
This fits in one line:
```suggestion
    def prepare_inputs_for_generation(self, decoder_input_ids, past, use_cache=None, **kwargs):
```

sgugger commented on tests/test_modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,793 @@
+        )
+
+        # Make sure `loss` exist
+        assert "loss" in outputs_encoder_decoder
```
```suggestion
        self.assertIn("loss", outputs_encoder_decoder)
```

sgugger commented on tests/test_modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,793 @@
+            model_2.save_pretrained(tmp_dirname)
+            model_1 = TFVisionEncoderDecoderModel.from_pretrained(tmp_dirname)
+
+            after_outputs = model_1(
+                pixel_values=pixel_values,
+                decoder_input_ids=decoder_input_ids,
+            )
```
```suggestion
            after_outputs = model_1(pixel_values=pixel_values, decoder_input_ids=decoder_input_ids)
```

sgugger reviewed (approved):
Thanks for adding this model!

ydshieh committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ydshieh committed:
Clean comments (need to work on TODOs for PyTorch models)

ydshieh commented:
@sgugger, thank you for your review, I have addressed them.

I am impressed by your ability to spot the `fits in one line` places.

I feel that sometimes `make style` can reformat in this case, but doesn't work well in a few places, as happened in this PR.

I dive a bit deeper, and found:

this one won't be reformatted by `make style`
```
    def prepare_inputs_for_generation(
        self,
        decoder_input_ids,
        past,
        use_cache=None,
        **kwargs,
    ):
```

but the following one will work well.

```
    def prepare_inputs_for_generation(
        self,
        decoder_input_ids,
        past,
        use_cache=None,
        **kwargs
    ):
```

The difference is the ending comma after the last argument. Is this a bug? I can open an issue if so.

ydshieh committed:
Remove TF -> PT in check_pt_tf_equivalence for TFVisionEncoderDecoderModel

ydshieh committed:
fixes

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -677,21 +682,29 @@ def generate(
-                output_hidden_states=output_hidden_states,
-                return_dict=return_dict_in_generate,
-            )
+            encoder_kwargs = {
```
fine with me!

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -677,21 +682,29 @@ def generate(
+        # (vision inputs might occur when the model is an encoder-decoder model)
         # Expand input ids if num_beams > 1 or num_return_sequences > 1
-        if num_return_sequences > 1 or num_beams > 1:
+        if len(shape_list(input_ids)) == 2 and (num_return_sequences > 1 or num_beams > 1):
```
that's a bit hacky - vision inputs should also work with `num_beams > 1` no? But ok for now until we do the big `generate` refactor @Rocketknight1

ydshieh replied:
Just as remark: The code inside this block treats `input_ids` as text-only inputs, for example `shape_list(input_ids)[-1]` assumes that the last dimension is the sequence dim.

For vision inputs, `generate` will be called only if it is a vision model as encoder in an encoder-decoder model (?). In this case, it is more the `decoder_input_ids` to be processed. And I think this is done in the next block

https://github.com/huggingface/transformers/blob/e2357599f7af1255b98729a3e0ef12bf2a081d44/src/transformers/generation_tf_utils.py#L722

It's not clear to me if an encoder-only vision model will need to call `generate()`. Maybe @NielsRogge can share some insights here (e.g. `ImageGPT`).

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -905,6 +906,16 @@ def generate(
                 input_ids, pad_token_id, eos_token_id, inputs_embeds
             )
 
+        if self.config.is_encoder_decoder:
```
can we remove this from the PR? This PR is for TF only no? I understand the need for this, but it's pretty hacky code that we should solve in a nicer way. I'll make the generate() function cleaner next week

patrickvonplaten commented on src/transformers/models/deit/modeling_deit.py:
```diff
@@ -477,7 +477,6 @@ class PreTrainedModel
     def forward(
         self,
         pixel_values=None,
-        attention_mask=None,
```
let's leave it for now - we can clean it up once the generate refactor is done: https://github.com/huggingface/transformers/issues/14421

patrickvonplaten commented on src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:
```diff
@@ -153,7 +153,7 @@
 class TFEncoderDecoderModel(TFPreTrainedModel):
     r"""
-    :class:`~transformers.TFEncoderDecoder` is a generic model class that will be instantiated as a transformer
+    :class:`~transformers.TFEncoderDecoderModel` is a generic model class that will be instantiated as a transformer
```
Thanks!

patrickvonplaten commented on tests/test_modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,768 @@
+# coding=utf-8
+# Copyright 2020 HuggingFace Inc. team.
```
```suggestion
# Copyright 2021- HuggingFace Inc. team.
```

patrickvonplaten commented on tests/test_modeling_vision_encoder_decoder.py:
```diff
@@ -64,14 +64,7 @@ def get_pretrained_model_and_inputs(self):
         pass
 
     def check_encoder_decoder_model_from_pretrained_configs(
-        self,
```
can we tackle all of this in a follow-up PR maybe? :-)

patrickvonplaten commented on tests/test_modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,768 @@
+        max_diff = np.max(np.abs(loss_pretrained - loss_init))
+        expected_diff = 0.0
+
+        self.assertAlmostEqual(max_diff, expected_diff, places=4)
```
Great tests! Could we add one more important test that shows that the architecture works as intended? 

E.g. could you simply copy paste your example here: https://huggingface.co/ydshieh/flax-vision-encoder-decoder-vit-gpt2-coco-en 

into a test so which a link to the image and an assert with the expected output. For this I think you need to upload a TF version of your model. 

It would be very nice to have it to be sure in the future that the model works correctly

patrickvonplaten reviewed (commented):
This is a great addition @ydshieh! Thanks a lot for the contribution. 

From my side it would be great if we could:
- remove all pytorch specific changes that are not needed to get the TF version working (I'll tackle this in a future PR :-) )
- add one slow test that ensures that the model works correctly

Thanks a bunch!

ydshieh commented:
> This is a great addition @ydshieh! Thanks a lot for the contribution.
> 
> From my side it would be great if we could:
> 
>     * remove all pytorch specific changes that are not needed to get the TF version working (I'll tackle this in a future PR :-) )
> 
>     * add one slow test that ensures that the model works correctly
> 
> 
> Thanks a bunch!

Hi @patrickvonplaten 

- OK for removing the changes on PT code.
- For the test, I think I forgot to do the same as I have done for PT/Flax

    https://github.com/huggingface/transformers/blob/8395f14de6068012787d83989c3627c3df6a252b/tests/test_modeling_vision_encoder_decoder.py#L666

    https://github.com/huggingface/transformers/blob/8395f14de6068012787d83989c3627c3df6a252b/tests/test_modeling_flax_vision_encoder_decoder.py#L467

    I will add it to TF test :)

ydshieh committed:
Revert changes in PT code.

ydshieh committed:
Update tests/test_modeling_tf_vision_encoder_decoder.py

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

ydshieh committed:
Add test_inference_coco_en for TF test

ydshieh committed:
fix quality

ydshieh committed:
fix name

ydshieh commented:
I removed the changes on PT code. I added 

```
class TFViT2GPT2ModelIntegrationTest(unittest.TestCase):
    @slow
    def test_inference_coco_en(self):
```
here
https://github.com/huggingface/transformers/blob/a179c4a5087e48127914e26e9895a455b42bac0d/tests/test_modeling_tf_vision_encoder_decoder.py#L775

ydshieh committed:
build doc

ydshieh committed:
add main_input_name

ydshieh committed:
Fix ckpt name in test

ydshieh committed:
fix diff between master and this PR

ydshieh committed:
fix doc

ydshieh committed:
Merge branch 'master' into tf_vision_encoder_decoder

# Conflicts:
#	docs/source/model_doc/auto.rst
#	src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py

ydshieh committed:
fix style and quality

ydshieh committed:
fix missing doc

ydshieh commented:
Hi, @patrickvonplaten @Rocketknight1 @NielsRogge 

I removed the changes on PT code. I also added 

```
class TFViT2GPT2ModelIntegrationTest(unittest.TestCase):
    @slow
    def test_inference_coco_en(self):
```
here
https://github.com/huggingface/transformers/blob/b3255863540fac38d49e5067c14fef7ae66dc31f/tests/test_modeling_tf_vision_encoder_decoder.py#L769

This PR is ready for review when you have the time :-)

ydshieh committed:
fix labels handling

patrickvonplaten commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,714 @@
+        return super().from_pretrained(pretrained_model_name_or_path, *model_args, **kwargs)
+
+    @classmethod
+    def from_encoder_decoder_pretrained(
```
We should probably add this part: https://github.com/huggingface/transformers/blob/8f6373c61ca71d2d8176e181b587a2307d8358c8/src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py#L402

here as well no? Otherwise it'll be difficult to load TF from PT

ydshieh replied:
Yes, great catch! For some reason I missed it in this PR. I will check again this PR against `TFEncoderDecoderModel` code.

ydshieh replied:
@patrickvonplaten  I corrected this part, and also updated the corresponding test (also forgot in the previous commits).

patrickvonplaten replied:
Thanks a lot!

patrickvonplaten commented on tests/test_modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,820 @@
+
+@require_vision
+@require_tf
+class TFViT2GPT2ModelIntegrationTest(unittest.TestCase):
```
Very nice!

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -628,14 +629,18 @@ def generate(
             bad_words_ids is None or isinstance(bad_words_ids, list) and isinstance(bad_words_ids[0], list)
         ), "`bad_words_ids` is either `None` or a list of lists of tokens that should not be generated"
 
+        # This block corresponds to the following line in `generation_utils`:
```
This whole function is in dire need of a refactor - I'll try to tackle this with @Rocketknight1 this month. Good for me the way it is not though

patrickvonplaten commented on src/transformers/generation_tf_utils.py:
```diff
@@ -691,21 +696,29 @@ def generate(
+                "return_dict": return_dict_in_generate,
+            }
+
+            # vision models don't use `attention_mask`.
```
nice!

patrickvonplaten reviewed (commented):
This looks more or less ready to be merged to me:

- I think we should add the loading `from_pt` hack as well to `from_encoder_decoder(...)` here
- I think we can delete the file: docs/source/model_doc/auto.rst no?

ydshieh commented:
> * auto.rst no?

Yes. I added the (empty) file to commit by mistake during git rebase/merge.

ydshieh committed:
Merge branch 'master' into tf_vision_encoder_decoder

ydshieh committed:
Delete auto.rst

ydshieh committed:
Add the changes done in #14016

ydshieh commented on tests/test_modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,824 @@
+
+            max_diff = np.max(np.abs(logits_tf_2.numpy() - logits_tf.numpy()))
+            self.assertAlmostEqual(max_diff, 0.0, places=3)
+
```
There was a `# TensorFlow => PyTorch` block, which did nothing (and if did, it would fail since we can't use `from_pretrained` along with `from_pt` or `from_tf` in this composite model.)

I removed it - need to remove the corresponding part in `test_modeling_tf_encoder_decoder.py` too in another PR.

patrickvonplaten commented:
@NielsRogge @Rocketknight1 - could you guys take a look here as well?

patrickvonplaten reviewed (approved):
Looks good to me now! I let @sgugger take a final look here

NielsRogge commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+            If `past_key_values` is used, optionally only the last `decoder_input_ids` have to be input (see
+            `past_key_values`).
+
+            Provide for sequence to sequence training to the decoder. Indices can be obtained using
+            [`PreTrainedTokenizer`]. See [`PreTrainedTokenizer.encode`] and [`PreTrainedTokenizer.__call__`] for
+            details.
```
The PyTorch model (`VisionEncoderDecoderModel`) automatically creates the decoder_input_ids by shifting the labels. Is this not the case for the TF one?

ydshieh replied:
No. I started this work before the new change (about `decoder_input_ids`) was done in `VisionEncoderDecoderModel`, and didn't follow it.

Would it be possible to leave it to another PR - I can make it, but prefer in a separate PR.

NielsRogge replied:
Ok makes sense!

NielsRogge replied:
See #14469

NielsRogge commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+    """
+    config_class = VisionEncoderDecoderConfig
+    base_model_prefix = "vision_encoder_decoder"
+    load_weight_prefix = "tf_vision_encoder_decoder_model_1"
```
Where does the _1 come from?

ydshieh replied:
I originally looked `TFRagModel` as reference for implementing the TF composite model, and saw

https://github.com/huggingface/transformers/blob/ac224bb0797c1ee6522d814139f3eb0a8947267b/tests/test_modeling_tf_rag.py#L973

I copied it and it works well (fixing the problems I had at that time), and I didn't think this part in more details.

I just spent some time on checking this again - and in fact, we can use `tf_vision_encoder_decoder_model` here, and `"tf_encoder_decoder_model` for `TFEncoderDecoderModel`.

Change this might break some user models - but since this is kind new models, and not popular yet, maybe it is worth the change. cc @patrickvonplaten , @Rocketknight1 , @sgugger, @LysandreJik for their thoughts on this.

(For `tf_rag_model_1`, maybe there is particular reason to make something work though.)

ydshieh replied:
I changed to "tf_vision_encoder_decoder_model" - better not to continue with the strange `_1`

ydshieh committed:
fix prefix

sgugger commented on src/transformers/models/encoder_decoder/modeling_tf_encoder_decoder.py:
```diff
@@ -148,8 +148,8 @@
-    the base model classes of the library as encoder and another one as decoder when created with the
+    [`TFEncoderDecoderModel`] is a generic model class that will be instantiated as a transformer architecture with one
+    of the base model classes of the library as encoder and another one as decoder when created with the
     :meth*~transformers.TFAutoModel.from_pretrained* class method for the encoder and
     :meth*~transformers.TFAutoModelForCausalLM.from_pretrained* class method for the decoder.
```
Those two lines need fixing
```suggestion
   [`~TFAutoModel.from_pretrained`] class method for the encoder and
   [`~TFAutoModelForCausalLM.from_pretrained`] class method for the decoder.
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+        kwargs: (*optional*) Remaining dictionary of keyword arguments. Keyword arguments come in two flavors:
+
+            - Without a prefix which will be input as `**encoder_kwargs` for the encoder forward function.
+            - With a *decoder_* prefix which will be input as `**decoder_kwargs`` for the decoder forward function.
```
```suggestion
            - With a *decoder_* prefix which will be input as `**decoder_kwargs` for the decoder forward function.
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+    r"""
+    [`TFVisionEncoderDecoderModel`] is a generic model class that will be instantiated as a transformer architecture
+    with one of the base vision model classes of the library as encoder and another one of the base model classes as
+    decoder when created with the :meth*~transformers.TFAutoModel.from_pretrained* class method for the encoder and
+    :meth*~transformers.TFAutoModelForCausalLM.from_pretrained* class method for the decoder.
```
```suggestion
    decoder when created with the [`~TFAutoModel.from_pretrained`] class method for the encoder and
    [`~TFAutoModelForCausalLM.from_pretrained`] class method for the decoder.
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
+        r"""
+        Initializing *TFVisionEncoderDecoderModel* from a pytorch checkpoint is not supported currently.
```
```suggestion
        Initializing `TFVisionEncoderDecoderModel` from a pytorch checkpoint is not supported currently.
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+
+
+        Params:
+            encoder_pretrained_model_name_or_path (:obj: *str*, *optional*):
```
```suggestion
            encoder_pretrained_model_name_or_path (`str`, *optional*):
```

sgugger commented on src/transformers/models/vision_encoder_decoder/modeling_tf_vision_encoder_decoder.py:
```diff
@@ -0,0 +1,731 @@
+                    - A path or url to a *pytorch index checkpoint file* (e.g, `./pt_model/`). In this case,
+                      `encoder_from_pt` should be set to `True`.
+
+            decoder_pretrained_model_name_or_path (:obj: *str*, *optional*, defaults to *None*):
```
```suggestion
            decoder_pretrained_model_name_or_path (`str`, *optional*, defaults to *None*):
```

sgugger reviewed (approved):
There are a few typos in the docstrings, but LGTM once it's fixed!
Thanks for all your work on this @ydshieh !

ydshieh committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ydshieh committed:
make style

ydshieh commented:
Applied @sgugger  review suggestions. Failed tests are unrelated.

sgugger commented:
Thanks again for all your work on this!
